### PR TITLE
Write up testing strategy

### PR DESCRIPTION
**Pull request by kevindew**
https://trello.com/c/wgkfOJsE/1283-improve-consistency-in-content-publisher-tests

We've struggled over time with a consistent approach to testing with
different understandings of our approaches. This document serves as an
attempt to put down our current thinkings. It's unlikely to be fully
exhaustive and I'm hoping it will be amended and improved over time.

I intend to follow this with a bunch of changes to our tests to increase
consistency.

[View rendered](https://github.com/alphagov/content-publisher/blob/test-strategy/docs/testing-strategy.md)

### CODE REVIEW

File: docs/testing-strategy.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+- they shouldn't test dead ends in a user flow (such as validation or
+  permission issues); these are simpler tested in
+  [request tests](#request-tests);
+- incidental behaviour (for example flash messages) shouldn't be tested, this
+  can be tested at a higher granularity in [unit](#unit-tests) or
+  [request](#request-tests) tests;
+- they shouldn't test the effects a test has on the database, only user visible
+  signs of success should be asserted on;
+- they shouldn't mock application code;
+- there may be mocking and assertions that particular API calls are made to
```
Comment (1pretz1) [inline docs/testing-strategy.md]: How about 'they may mock and/or assert that particular API calls are made to external services;', it follows the wording flow of line 65 which I think is an improvement - https://github.com/alphagov/content-publisher/pull/1590/commits/b55b02737011b70e64629e2090964458c3b96158#diff-18a0a75d23cfa0c005b7d73b8c2477ddR65.

File: docs/testing-strategy.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+tests in that they test a single endpoint at a time and can exhaust the logical
+outcomes of that endpoint where some situations may be difficult for
+a user to fall into.
+
+Characteristics of request tests:
+
+- they should test the variety of responses an endpoint returns, checking
+  aspects of the response such as status code and any flash messages;
+- they shouldn't test responses where there isn't specific logic written for
+  the scenario;
```
Comment (1pretz1) [inline docs/testing-strategy.md]: This line doesn't really mean much to me, does this basically mean that we shouldn't use request tests if the controller action is simple and doesn't house any 'complex' logic?

File: docs/testing-strategy.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+- they should assert against the HTTP responses involved in the request, and
+  not side-effects such as database changes, these are better suited to [unit
```
Comment (1pretz1) [inline docs/testing-strategy.md]: ➕

File: docs/testing-strategy.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+- they should focus on the effects of a single endpoint with a single HTTP
+  method at a time, for example `POST /documents`, for multiple endpoints
+  consider a [feature test](#feature-tests);
+- they act as integration tests and thus shouldn't mock application code;
+- they can mock external API calls, but these should not be asserted against -
+  this is better suited to [unit tests](#unit-tests);
+- they shouldn't be used to extensively test HTML output, this is better suited
+  by [view tests](#view-tests);
+- they may not be necessary for places where a response doesn't have logic
+  and already has coverage provided by a feature test.
```
Comment (1pretz1) [inline docs/testing-strategy.md]: This brings up a question in my head on whether we should test in request tests the 'happy path' which is more than likely covered by a feature test. We are currently doing this for a few of the requests tests (withdraw and unwithdraw etc) but should this be a hard and fast rule?

File: docs/testing-strategy.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
+# Testing Strategy
+
+We aim for the Content Publisher to have a test suite that provides developers
```
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: Thanks - missed that in my proof read 🤦‍♂

File: docs/testing-strategy.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+- pragmatism in what we test - if all code paths of a class/module are
+  exhausted by other tests then we may not need to explicitly test something
+  further.
+
+This testing strategy was inspired by:
+[Whitehall Testing Guidelines][whitehall-tests],
+[Email Alert API Testing][email-alert-api-tests], and
+[Thoughtbots' How We Test Rails Applications][thoughtbot-rails-tests]
+
+## Test organisation
```
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: I'll go with "Tests directory structure" - does that work for you?

File: docs/testing-strategy.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+  alternative;
+- features - contains [feature tests](#feature-tests), which exercises
+  application features from a user perspective;
+- javascripts - contains [unit tests](#unit-tests) for JavaScript files
+  and configuration and helper files for Jasmine;
+- requests - contains [request tests](#request-tests), which test the HTTP API
+  of the application;
+- support - contains helper files for shared test methods and logic;
+- views - contains [view tests](#view-tests), which are used to test
+  [Rails views][rails-views] output expected HTML in a given scenario.
```
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: Oh good catch 👍

File: docs/testing-strategy.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+  to the database or making API calls.
+
+## Feature tests
+
+Feature tests are used to assert that Content Publisher behaves correctly from
+the perspective of someone using the application. This ensures we're testing
+the system in a way representative of expected user behaviour. This makes the
+tests some of the most valuable of the application, however they are slower
+than other tests to run and can be difficult to debug. Therefore they are not
+intended to exhaustively test all the scenarios that can occur.
```
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: Hopefully this should be covered with: https://github.com/alphagov/content-publisher/blame/b55b02737011b70e64629e2090964458c3b96158/docs/testing-strategy.md#L86-L88

I wanted to avoid the phrase "happy path" as I think it's a bit more nuanced than that in that if you have a long negative flow that is a feature in itself like scheduled publishing failed it doesn't quite make sense.

File: docs/testing-strategy.md [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+Characteristics of feature tests:
+
+- they should test across multiple requests - consider a
+  [request test](#request-tests) to test a single request;
```
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: Sure, I got a bit hung up on the redirect part of request tests as they do generate a second request but it tests the integration from the first one. So I'm thinking of them like a forwarding rather than multiple requests. Hmm I'll have a think about if this can be rephrased - and same for similar section in request tests.

File: docs/testing-strategy.md [hunk 10]
```diff
@@ -0,0 +1,152 @@
+# Testing Strategy
+
+We aim for the Content Publisher to have a test suite that provides developers
+with value. We want the application to be fast to test; for the tests to catch
+breaking behavioural changes; and for it to be easy for developers to know
+where to add tests. The purpose of this document is
```
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: I don't think we're explicitly aiming for maximum test coverage (that'd be 100%) as that seems to often be an anti-pattern to be too worried about that number.

File: docs/testing-strategy.md (hunk 1 above)
Comment (kevindew) [reply docs/testing-strategy.md]: Good shout.

File: docs/testing-strategy.md (hunk 2 above)
Comment (kevindew) [reply docs/testing-strategy.md]: I meant it more about where Rails may do things automatically. In that we probably don't want request tests that test whether we get a 404 for every document that doesn't exist or a 400 when required parameters are missing.

I wanted it to be a little more nuanced that just don't test Rails exceptions as sometimes we'd want to test them when our code interacts with them - for example we might want to test we get a 404 when we try access an image for a document where the image exists but belongs to a different document.

File: docs/testing-strategy.md (hunk 4 above)
Comment (kevindew) [reply docs/testing-strategy.md]: I'd think if a route has different responses it should have request tests. It doesn't sit quite right with me to skip one of them as that would be covered by a feature test (feels overly coupled).

Just seems a bit overkill to test very basic requests which are just show something. Does that make sense? Happy to try rephrase a bit if you have some ideas.

**Comment by kevindew** [comment]
> Thanks for writing this up, looks great so far! Just left a few comments on slight improvements / general questions.

Thanks for the review Pete 👍 

> I like the addition of request testing to bridge the gap between feature and unit testing and also would like to see more testing of our interactors as there are still quite a few that are untested, what do you think?

Yeah I'm hoping we gradually get tests for all of them.

> Also just to make sure of my understanding of when to use a 'view' test, should these be used when a partial contains complex logic on that HTML it displays depending on what's being passed in?

Yeah I think anytime you want to be testing that certain HTML is output a view test is the right way to go. I'm hoping https://github.com/alphagov/content-publisher/pull/1592 serves as a reasonably good example where we were using feature tests to test stuff that is isolated to a particular view.

File: docs/testing-strategy.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (1pretz1) [reply docs/testing-strategy.md]: Hmm true, I think my wording was a lil dodgy here. I was more meaning that all of the features we build are thoroughly tested rather than worrying that every line of code has been hit. However, we're apparently hitting 97.74% code coverage according to RSpec's report which sounds nice! 😄

File: docs/testing-strategy.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (1pretz1) [reply docs/testing-strategy.md]: Ah, you have a point. 👍

File: docs/testing-strategy.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (1pretz1) [reply docs/testing-strategy.md]: Yup looks good to me!

File: docs/testing-strategy.md (hunk 4 above)
Comment (1pretz1) [reply docs/testing-strategy.md]: That makes sense, I think I was coming at it from the point of if I'm creating a request test for a thing which can have multiple responses should I always cover the scenario in which the feature covers ('happy' scenario). This is answered by your first paragraph! 😄

File: docs/testing-strategy.md (hunk 2 above)
Comment (1pretz1) [reply docs/testing-strategy.md]: Oh okay I get you, maybe add an example like the ones you've stated to the bullet point? As without that context I feel it's quite hate to decipher what it actually means, at least to my brain! 😆

File: docs/testing-strategy.md [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+The principles we most value in testing are:
+
+- consistency across tests over developer preference - all tests should
+  feel like they could have been written by the same person;
+- readable tests that are easy to follow - aim for expectations to be clear,
+  setup code to be close to associated assertions and avoid being overzealous
+  in removing repetition;
+- avoid testing configuration - tests should be focused on testing logic we've
+  introduced, rather than whether or not our code was typed correctly;
+- pragmatism in what we test - if all code paths of a class/module are
```
Comment (leenagupte) [inline docs/testing-strategy.md]: I agree with this, but I think it would be helpful if existing tests were sign posted somewhere, like the commit message?

File: docs/testing-strategy.md [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+The purpose of view tests is to assert that the expected HTML is rendered in a
+particular scenario. Typically it isn't necessary to have view tests as the
+use of complex logic is discouraged in views and often aspects of views are
+implicitly tested in feature or request tests. However when it is suited for
+us to test particular HTML output they are the most appropriate choice.
+
+Characteristics of view tests:
+
+- they should make assertions based on logic in the view and not be used to
+  assert that HTML is as we wrote it;
```
Comment (ChrisBAshton) [inline docs/testing-strategy.md]: Could you elaborate on this line? Based on https://github.com/alphagov/content-publisher/pull/1592, I'd be tempted to add:

```suggestion
  assert that HTML is as we wrote it (for example, assert that a selector contains specific text,
  rather than testing for the presence of all classes, IDs and data attributes a node might have);
```

File: docs/testing-strategy.md [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+[RSpec Rails view tests][rspec-rails-views].
+
+[whitehall-tests]: https://github.com/alphagov/whitehall/blob/099f53e35cdd0ea63a2349be3766c98e65521ce3/docs/testing.md
+[email-alert-api-tests]: https://github.com/alphagov/email-alert-api/pull/300/commits/bf087c28442629bf672be67441fd224f6d746fa5
+[thoughtbot-rails-tests]: https://thoughtbot.com/blog/how-we-test-rails-applications
+[factory-bot]: https://github.com/thoughtbot/factory_bot
+[rails-views]: https://guides.rubyonrails.org/testing.html#testing-views
+[readable-feature-tests]: https://about.futurelearn.com/blog/how-we-write-readable-feature-tests-with-rspec
+[interactors]: adr/0011-moving-business-logic-out-of-controllers.md
+[rspec-rails-views]: https://relishapp.com/rspec/rspec-rails/v/3-9/docs/view-specs/view-spec
```
Comment (ChrisBAshton) [inline docs/testing-strategy.md]: Could you make this list of link texts alphabetical?

File: docs/testing-strategy.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: I've added a section in request tests to cover this now.

File: docs/testing-strategy.md (hunk 2 above)
Comment (kevindew) [reply docs/testing-strategy.md]: Sure - I've added that in, hopefully it's worded ok wasn't as easy to phrase as I thought.

File: docs/testing-strategy.md (hunk 11 above)
Comment (kevindew) [reply docs/testing-strategy.md]: I wouldn't say that this is a concern as part of test strategy (as I don't think we've pushed people to do so) so I don't think it belongs here. It sounds like a good courtesy though and be interesting to keep an eye out for if we think we have a problem in this regard.

File: docs/testing-strategy.md (hunk 13 above)
Comment (kevindew) [reply docs/testing-strategy.md]: I'm not aware that we have a convention for the ordering of links in markdown. I've gone with ordering them based on the order they appear in the doc.

File: docs/testing-strategy.md [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+those within the app directory. Code that is stored in the lib directory has
+corresponding tests in the spec/lib directory.
+
+Directories within spec that don't contain Ruby unit tests are as follows:
+
+- factories - contains [FactoryBot factories][factory-bot], which are a fixture
+  alternative;
+- fixtures - contains supporting files for tests (for example image files) which
+  aren't easily produced by factories;
+- features - contains [feature tests](#feature-tests), which exercises
```
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: Sure - the subjective nature of the word itself makes it hard to define definitively.

My definition of feature is often based around what you'd use to advertise what the app can do. In other words, what distinguishes it. I think that is somewhat similar to your phrasing, however I'd want to tweak that slightly to be: a task that a user accomplishes through the application. I think that reduces the broadness a bit and I've gone with that in my last edit

File: docs/testing-strategy.md [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+- views - contains [view tests](#view-tests), which are used to test
+  [Rails views][rails-views] output expected HTML in a given scenario.
+
+## Unit tests
+
+The purpose of these tests is to test individual units of the system in
+isolation. They are intended to provide exhaustive tests of the code paths
+through a particular class or module through its public interface. As per
+the [test pyramid](https://martinfowler.com/bliki/TestPyramid.html) approach
+to software testing, these should provide the majority of test coverage.
```
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: Maybe we're on a slightly different wavelength here. I can see this is somewhat ambiguous based on how the word coverage is understood. I've amended this to instead reference the volume of tests.

File: docs/testing-strategy.md [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+Characteristics of feature tests:
+
+- they should test across multiple requests - consider a
+  [request test](#request-tests) to test a single request;
+- they should navigate the application using the same means as a user, for
+  example clicking links or submitting forms;
+- they shouldn't test dead ends in a user flow (such as validation or
+  permission issues); these are simpler tested in
+  [request tests](#request-tests);
+- incidental behaviour (for example flash messages) shouldn't be tested, this
```
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: I've decided to remove this line as it seems to be an outlier that doesn't quite represent current/past patterns - it's somewhat borrowed from the Whitehall guidelines as an approach I agree with but is not in practice by us sufficiently.

To explain my perspective, I'd suggest that you want the feature test to assert that a user has completed their task in one way and not every visual side effect of the change - which would be under the definition of incidental. Hopefully when a user has performed a task there is an indication of their change on the outcome page that can be asserted against (e.g. a change in data). 

I guess as an example of something incidental I'd pick [content data link in publish test](https://github.com/alphagov/content-publisher/blob/master/spec/features/workflow/publish_spec.rb#L57-L61). It's something that happened because I published, but it doesn't only happen on publish so should every state test it, nor does it feel a place you'd expect to change a test if you were changing the content data admin link.

File: docs/testing-strategy.md [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+  and not test logic defined elsewhere;
+- they may mock dependent objects and/or assert that particular external
+  methods are called;
+- they should test all code paths through a class/module;
+- they should test that side effects intended by the code occur, for example writing
+  to the database or making API calls.
+
+## Feature tests
+
+Feature tests are used to assert that Content Publisher behaves correctly from
```
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: Thanks, I've given this section a rephrasing.

File: docs/testing-strategy.md [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+outcomes of that endpoint where some situations may be difficult for
+a user to fall into.
+
+Characteristics of request tests:
+
+- they should test the variety of responses an endpoint returns, checking
+  aspects of the response such as status code and any flash messages;
+- they shouldn't test responses where there isn't specific logic written for
+  the scenario;
+- they should assert against the HTTP responses involved in the request, and
```
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: I'm not sure I quite understand, could you rephrase or add more context?

File: docs/testing-strategy.md [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+  not side-effects such as database changes, these are better suited to [unit
+  testing](#unit-tests) an [interactor][interactors];
+- they should focus on the effects of a single endpoint with a single HTTP
+  method at a time, for example `POST /documents`, for multiple endpoints
+  consider a [feature test](#feature-tests);
+- they act as integration tests and thus shouldn't mock application code;
+- they can mock external API calls, but these should not be asserted against -
+  this is better suited to [unit tests](#unit-tests);
+- they shouldn't be used to extensively test HTML output, this is better suited
```
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: Sure I've applied a change for that

**Comment by kevindew** [comment]
Thanks for the reviews, I've made a number of amends based off them.

> @ChrisBAshton 
> * Do we want to set a target value for our "fastness of tests" metric? Have some automated build steps in place which measures time taken to run the test suite and fails if it makes the suite significantly slower?
> * Is this a good opportunity to improve the workflow of writing our tests (as opposed to what they look like at the end)? I'm aware some people run their tests inside Docker, some outside. Some run `rake`, others run more specific commands. Linting errors often get missed if the `rake` command isn't used, or if it fails before it gets to the linting stage. Linting is also much faster than running the entire test suite - would it be worth putting it first in the `rake` tasks queue so that we fail faster?

Both of those things seem like interesting areas but likely I'd consider them to be outside the scope of what this document is trying to achieve.

In terms of speed it is good to be trying to have fast tests. I think it would be a challenge to have such a tool that identifies slow tests and calls them out, as I think this would be an area that is very difficult to have a consistent rule. I'm also hoping that in the not too distant future RSpec-Rails embraces parallelism like Rails 6 did and thus provide speed boost (could be a while though as they started talking about parallel RSpec 6 years ago)

In terms of the rake ordering and approach to run tests, I think mostly it's somewhat of a different strokes. Myself I mostly run individual tests with RSpec directly and then rake on its own occasionally to check everything together. I'm not aware of a particular problem that warrants widening the scope of this to include that but we can certainly chat about it if there are challenges you are facing.

> Thanks for putting this together @kevindew! I like the consistency it will bring, and we definitely have enough tests to allow us to reason about the different alternatives. There are a couple of comments that I'd like to clarify before this goes in, the most interesting one being around the use of feature tests.
> 
> Otherwise, in looking at the different alternatives I noticed one that's missing: "integration tests". This is probably intentional, but it might be helpful to briefly state what we're not doing as well.

Sure, it's not something particular intentional - in the same way we're not calling out functional tests explicitly. I wouldn't say we've got a firm stance on integration tests myself (as I recall them briefly existing for a few days last year) just that request and feature tests cover our current needs for integration and functional testing but were they to not be the right way to test a part of the app (which may occur with migration) then we'd consider having directories for them. I'm not too keen on wanting to add that to the doc, more it feels like a reason to amend the doc were our needs to change.

File: docs/testing-strategy.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (1pretz1) [reply docs/testing-strategy.md]: Looks good to me! ⭐️

File: docs/testing-strategy.md (hunk 2 above)
Comment (1pretz1) [reply docs/testing-strategy.md]: Had a read through and it's much clearer! 🌟

File: docs/testing-strategy.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply docs/testing-strategy.md]: On reflection, I think it's fine. The example I had in mind was contrived - lets wait and see.

File: docs/testing-strategy.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply docs/testing-strategy.md]: Thanks for explaining. Leaving it out/flexible sounds good to me, although I like your example of the link to Content Data, as that's definitely not part of the "publish feature".

File: docs/testing-strategy.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply docs/testing-strategy.md]: Thanks for rephrasing - that is clearer. Perhaps it's just me, but when I read this I get the impression that I "should" write unit tests for everything so that we're "following" the pyramid model. Looking at it again: what point are we trying to communicate in this sentence?

**Comment by kevindew** [comment]
Thanks @benthorner I've merged in to move forward, but happy to carry on talking about the unit test point and update accordingly - after all I hope this doc evolves with us.

File: docs/testing-strategy.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: Sure - I see what you mean and writing unit tests for "everything" is frequently impractical or lacks pragmatism. More so that unit tests should be the primary mechanism to test logic that we've wrote, and we should be using that to explore the distinct logical paths that code takes thus producing the volume of tests. Since this offers advantages in terms of ease of authoring, ease of debugging and speed of execution compared to higher level tests.

I guess the part of the testing pyramid part description from Martin Fowler that resonates in reference to this is:

"I always argue that high-level tests are there as a second line of test defense. If you get a failure in a high level test, not just do you have a bug in your functional code, you also have a missing or incorrect unit test. Thus I advise that before fixing a bug exposed by a high level test, you should replicate the bug with a unit test. Then the unit test ensures the bug stays dead."

[the remark this answers is not in the document]
Comment (benthorner) [reply docs/testing-strategy.md]: Oh dear, I find myself disagreeing with this last paragraph even more: a failure in a high level test doesn't doesn't to me suggest a test is missing, since the existing tests are clearly serving their purpose. The argument is very strange.

I did wonder about simply suggesting "will provide" instead of "should provide", but that's getting too subtle and subjective.

I think we're both in agreement about the pros and cons of (starting with) unit testing. For me, the important question is: what is the *easiest* way to test this? If there's no unit test for something, then writing a full one may be *harder* and possibly more resource intensive than simply modifying a higher-level test.

Let's wait and see if the way I'm thinking about tests rubs against the strategy. Then we can have a more focussed discussion.

[the remark this answers is not in the document]
Comment (kevindew) [reply docs/testing-strategy.md]: ha - oh dear indeed. 

I'd suggest that the high level test is serving its purpose in telling you that the functionality is not working however the nature of high level tests shouldn't be telling you which unit of code is not working so you have a gap. If you're lucky enough to have an exception be the problem then you can work out the unit easily enough but otherwise you're digging into UI assertions.

But yes, let us chat more if this becomes actual pain points.
